lib/referencer.js

```diff
@@ -453,6 +459,17 @@ class Referencer extends esrecurse.Visitor {
         this.visitProperty(node);
     }
 
+    PropertyDefinition(node) {
+        if (node.computed) {
+            this.visit(node.key);
+        }
+        if (node.value) {
+            this.scopeManager.__nestFunctionScope(node, /* strict= */ true);
```

Comment:
how come you are nesting a function scope specifically for a property value?
or rather - why do you nest a scope at all for properties?
They can't declare variables (unless ofc the value is a function), right?

Reply:
In the spec, class field initializers are implicit functions. Those implicit functions are called from the class instantiation process. This scope represents that straightly.

The expressions in the computed properties are evaluated one-time at defining the class. On the other hand, the expressions in the class field initializers are evaluated multiple times at instantiating class instances. I think it's odd if those are mixed in the same scope.

> They can't declare variables (unless ofc the value is a function), right?

The access to the `arguments` variable in the class field initializers is a syntax error, so we can't access the local variable of the implicit functions. But I'm not sure how the direct `eval()` behaves in the implicit functions.

Reply:
Interesting!

Relevant part of the spec (I think):
In particular 3.d.

https://tc39.es/proposal-class-fields/#runtime-semantics-class-field-definition-evaluation

```
3. If Initializer_opt is present,
    a. Let lex be the Lexical Environment of the running execution context.
    b. Let formalParameterList be an instance of the production FormalParameters:[empty] .
    c. Let privateScope be the PrivateEnvironment of the running execution context.
    d. Let initializer be FunctionCreate(Method, formalParameterList, Initializer, lex, true, privateScope).
    e. Perform MakeMethod(initializer, homeObject).
    f. Let isAnonymousFunctionDefinition be IsAnonymousFunctionDefinition(Initializer).
```

If I'm understanding it correctly that means that it's essentially saying that at runtime it essentially defines a field as this:

```js
class Foo {
  prop = 1;
}
```

Is more or less the same as

```js
class Foo {
  constructor() {
    this.prop = (() => 1)();
  }
}
```

With the one caveat that if the intialiser contains the identifier `arguments`, then it's a syntax error (as defined in [section 2.17.2](https://tc39.es/proposal-class-fields/#sec-performeval-rules-in-initializer))